File: 0142-linked-list-cycle-ii/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* slow_ptr = head;
        ListNode* fast_ptr = head;
        ListNode* intersect = NULL;
        
        // Start from head.
        while (true)
        {
            // If no cycle, fast_ptr will reach the end first.
            if ((fast_ptr == NULL) || (fast_ptr->next == NULL))
            {
                return NULL;
            }
            
            // Move two steps at a time.
            fast_ptr = fast_ptr->next->next;
            
            // Move one step at a time.
            slow_ptr = slow_ptr->next;
            
            // A cycle exists if two pointers meet.
            if (fast_ptr == slow_ptr)
            {
                intersect = slow_ptr;
                break;
            }
        }
        
        ListNode* ptr1 = head;
        ListNode* ptr2 = intersect;
        while (ptr1 != ptr2){
            ptr1 = ptr1->next;
            ptr2 = ptr2->next;
        }
        return ptr1;
    }
};

File: 0142-linked-list-cycle-ii/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return
        slow = head
        fast = head
        while True:
            if not fast.next:
                return
            fast = fast.next.next
            slow = slow.next
            if not fast or not slow:
                return
            if slow == fast:
                break
        fast = head
        while True:
            if slow == fast:
                return slow
            slow = slow.next
            fast = fast.next

